src/pyrokinetics/equilibrium/eliteinp.py

```diff
@@ -0,0 +1,212 @@
+
+from ..constants import electron_charge
+from ..file_utils import FileReader
+from ..typing import PathLike
+from ..units import UnitSpline
+from ..units import ureg as units
+from .equilibrium import Equilibrium
+
+
+def read_eqin(filename_or_file):
+    def token_generator(fp):
+        for line in fp:
+            for tok in line.split():
+                yield tok
+
+    with open(filename_or_file) as f:
+        f.readline()
+        tokens = token_generator(f)
+
+        try:
+            npsi, npol = int(next(tokens)), int(next(tokens))
+        except (StopIteration, ValueError):
+            raise IOError("Second line should contain Npsi and Npol")
+
+        data_dict = {"npsi": npsi, "npol": npol}
+
+        while True:
+            try:
+                varname = next(tokens).rstrip(":")
+            except StopIteration:
+                break
+
+            try:
+                if varname.lower() in {"r", "z"} or varname.startswith("B"):
+                    data = np.array([float(next(tokens)) for _ in range(npsi * npol)])
+                    data = data.reshape((npsi, npol), order="F")
+                elif varname in {"Zeff", "Zimp", "Aimp", "Amain"}:
+                    data = float(next(tokens))
+                else:
+                    data = np.array([float(next(tokens)) for _ in range(npsi)])
+            except (StopIteration, ValueError):
+                raise IOError(f"Error while reading {varname}")
+
+            data_dict[varname] = data
+
+    necessary_keys = ["Psi", "fpol", "q", "R", "z"]
+    if not all(key in data_dict.keys() for key in necessary_keys):
+        raise ValueError(
+            "Missing equilibrium data in ELITEINP file."
+            " Currently only HELENA ELITEINP files are supported, please"
+            " raise issue if another source (like) SCENE is needed"
+        )
+
+    if "Psi" in data_dict:
+        data_dict["Psi"] = data_dict["Psi"] * units.weber / units.radian
+
+    if "dp/dpsi" in data_dict:
+        data_dict["p_prime"] = (
+            data_dict["dp/dpsi"] * units.pascal / (units.weber / units.radian)
+        )
+
+    if "ffp" in data_dict:
+        data_dict["ff_prime"] = data_dict["ffp"] * (
+            units.tesla**2 * units.meter**2 * units.radian / units.weber
+        )
+
+    if "Bt" not in data_dict and "fpol" in data_dict and "R" in data_dict:
+        fpol = data_dict["fpol"]
+        R = data_dict["R"]  # shape (npsi, npol)
+        data_dict["Bt"] = fpol[:, None] * R  # shape (npsi, npol)
+    else:
+        raise ValueError("Missing Bt data in ELITEINP file")
+
+    if "p" not in data_dict:
+        if "ne" in data_dict and "Te" in data_dict:
+            data_dict["p"] = (
+                data_dict["ne"] * (data_dict["Te"] * electron_charge.m)
+            ) * units.pascal
+        else:
+            raise ValueError(
+                "No electron density and temperature data found in ELITEINP"
+            )
```

Comment:
I usually recommend flipping checks like this

```suggestion
        if "ne" not in data_dict or "Te" not in data_dict:
            raise ValueError(
                "No electron density/temperature data found in ELITEINP"
            )

        data_dict["p"] = (
            data_dict["ne"] * (data_dict["Te"] * electron_charge.m)
        ) * units.pascal
```

Reply:
Done